remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -422,75 +428,134 @@ impl REClient {
 
     pub async fn upload(
         &self,
-        metadata: RemoteExecutionMetadata,
+        _metadata: RemoteExecutionMetadata,
         request: UploadRequest,
     ) -> anyhow::Result<UploadResponse> {
-        let mut client = self.grpc_clients.cas_client.clone();
-
-        let files_with_digest: Vec<Request> = request
-            .files_with_digest
-            .unwrap_or_default()
-            .into_try_map(|x| {
-                anyhow::Ok(Request {
-                    digest: Some(tdigest_to(x.digest)),
-                    // FIXME: This could do a lot of blocking reads
-                    data: fs_util::read(&x.name)?,
-                    compressor: compressor::Value::Identity as i32,
-                })
-            })?;
-
-        let re_request = BatchUpdateBlobsRequest {
-            instance_name: INSTANCE_NAME.into(),
-            requests: [
-                request
-                    .inlined_blobs_with_digest
-                    .unwrap_or_default()
-                    .into_map(|x| Request {
-                        digest: Some(tdigest_to(x.digest)),
-                        data: x.blob,
-                        compressor: compressor::Value::Identity as i32,
-                    }),
-                files_with_digest,
-            ]
-            .concat(),
+        let mut bs_client = self.grpc_clients.bytestream_client.clone();
+
+        /* 4MB, with some slop just in case */
+        /* XXX FIXME (aseipp): can this come from grpc transport? CacheCapabilities? */
+        let grpc_msg_limit = (4 * 1000 * 1000) - 4096;
+
+        let mut blob_hashes = vec![];
+
+        let instance_prefix = if let Some(name) = self.instance_name.clone() {
+            format!("{}/", name)
+        } else {
+            "".to_string()
         };
 
-        let blob_hashes = re_request
-            .requests
-            .iter()
-            .map(|x| x.digest.as_ref().unwrap().hash.clone())
-            .collect::<Vec<String>>();
-        let response = client
-            .batch_update_blobs(with_internal_metadata(re_request, metadata))
-            .await?;
+        // upload inlined blobs
+        let inlined_blobs = request.inlined_blobs_with_digest;
+        for blob in inlined_blobs.unwrap_or_default() {
```

Comment:
This has got no concurrency, so we upload all the files in sequence. Can you add that in? It should be a matter of putting all those futures into a Vec, then using `futures::future::try_join_all` on it to execute them concurrently?

You'll need to clone the `bs_client` to be able to use it in all those futures (because it wants  `&mut self`), but that's fine, that's how you're supposed to use it. 

(in an ideal world we would probably have some limits on concurrency via the use of a semaphore, but that can come later)